File: src/agent_search_1688/mcp_server.py
"""将 1688 Agent Search 工具注册表暴露为 stdio MCP Server。"""

from __future__ import annotations

import json
import sys
from typing import Any

from .tools import ToolRegistry
from .web_search_tool import build_1688_tool_registry


MCP_PROTOCOL_VERSION = "2024-11-05"


def _response(request_id: Any, result: dict[str, Any]) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": request_id, "result": result}


def _error(request_id: Any, code: int, message: str) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": request_id, "error": {"code": code, "message": message}}


def handle_mcp_message(message: dict[str, Any], registry: ToolRegistry) -> dict[str, Any] | None:
    method = message.get("method")
    request_id = message.get("id")
    if method == "notifications/initialized":
        return None
    if method == "initialize":
        return _response(
            request_id,
            {
                "protocolVersion": MCP_PROTOCOL_VERSION,
                "capabilities": {"tools": {}},
                "serverInfo": {"name": "1688-tools", "version": "0.3.0"},
            },
        )
    if method == "tools/list":
        return _response(request_id, {"tools": registry.definitions()})
    if method == "tools/call":
        params = message.get("params")
        if not isinstance(params, dict):
            return _error(request_id, -32602, "tools/call 参数无效")
        name = params.get("name")
        arguments = params.get("arguments", {})
        if not isinstance(name, str) or not isinstance(arguments, dict):
            return _error(request_id, -32602, "工具名称或参数无效")
        try:
            result = registry.dispatch(name, arguments)
        except (KeyError, ValueError) as exc:
            return _response(request_id, {"content": [{"type": "text", "text": str(exc)}], "isError": True})
        except Exception:
            return _response(request_id, {"content": [{"type": "text", "text": "web_search 执行失败"}], "isError": True})
        return _response(
            request_id,
            {"content": [{"type": "text", "text": json.dumps(result, ensure_ascii=False)}]},
        )
    return _error(request_id, -32601, f"不支持的 MCP 方法：{method}")


def run_1688_mcp_server() -> int:
    registry = build_1688_tool_registry()
    for raw_line in sys.stdin:
        try:
            message = json.loads(raw_line)
        except json.JSONDecodeError:
            continue
        if not isinstance(message, dict):
            continue
        response = handle_mcp_message(message, registry)
        if response is not None:
            print(json.dumps(response, ensure_ascii=False), flush=True)
    return 0

File: src/tools/mcp-server.ts
import { createInterface } from "node:readline";

import { resolveSkillRoot } from "../config.js";
import { ToolRegistry } from "./registry.js";
import { buildToolRegistry } from "./web/search.js";

export const MCP_PROTOCOL_VERSION = "2024-11-05";

export async function handleMcpMessage(
  message: Record<string, unknown>,
  registry: ToolRegistry,
): Promise<Record<string, unknown> | undefined> {
  const method = message.method;
  const id = message.id;
  if (method === "notifications/initialized") return undefined;
  if (method === "initialize") {
    return response(id, {
      protocolVersion: MCP_PROTOCOL_VERSION,
      capabilities: { tools: {} },
      serverInfo: { name: "1688-tools", version: "1.0.0" },
    });
  }
  if (method === "tools/list") return response(id, { tools: registry.definitions() });
  if (method === "tools/call") {
    if (!isRecord(message.params)) return rpcError(id, -32602, "tools/call 参数无效");
    const name = message.params.name;
    const arguments_ = message.params.arguments ?? {};
    if (typeof name !== "string" || !isRecord(arguments_)) return rpcError(id, -32602, "工具名称或参数无效");
    try {
      const result = await registry.dispatch(name, arguments_);
      return response(id, { content: [{ type: "text", text: JSON.stringify(result) }] });
    } catch (error) {
      const detail = error instanceof Error ? error.message : String(error);
      return response(id, {
        content: [{ type: "text", text: detail.slice(0, 1_000) }],
        isError: true,
      });
    }
  }
  return rpcError(id, -32601, `不支持的 MCP 方法：${String(method)}`);
}

export async function runMcpServer(options: {
  readonly input?: NodeJS.ReadableStream;
  readonly output?: { write(chunk: string): unknown };
  readonly cwd?: string;
} = {}): Promise<number> {
  const input = options.input ?? process.stdin;
  const output = options.output ?? process.stdout;
  const registry = buildToolRegistry({ skillRoot: resolveSkillRoot(options.cwd ?? process.cwd()) });
  try {
    const lines = createInterface({ input });
    for await (const line of lines) {
      let message: unknown;
      try { message = JSON.parse(line); } catch { continue; }
      if (!isRecord(message)) continue;
      const result = await handleMcpMessage(message, registry);
      if (result) output.write(`${JSON.stringify(result)}\n`);
    }
    return 0;
  } finally {
    await registry.close();
  }
}

function response(id: unknown, result: Record<string, unknown>): Record<string, unknown> {
  return { jsonrpc: "2.0", id, result };
}

function rpcError(id: unknown, code: number, message: string): Record<string, unknown> {
  return { jsonrpc: "2.0", id, error: { code, message } };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
